numba/types/npytypes.py

```diff
@@ -265,6 +265,10 @@ class Array(Buffer):
 
     def __init__(self, dtype, ndim, layout, readonly=False, name=None,
                  aligned=True):
+        if isinstance(dtype, DType):
```

Comment:
That doesn't really sound right. Perhaps we could do this in `__getitem__` where we handle the `dtype[:]` syntax, but I don't think we should let pass internal bugs where we give the wrong dtype when constructing an Array type instance.